src/plugin/dl/dlwrappers.cpp

```diff
@@ -136,15 +136,22 @@ dlopen_try_paths(const char *filename, int flag, string path)
 extern "C"
 void *dlopen(const char *filename, int flag)
 {
+  // TODO(kapil): Replace with scoped lock once we have it.
   bool lockAcquired = dmtcp_libdlLockLock();
   string rpath;
   string runpath;
   void *ret = NULL;
 
   getRpathRunPath(__builtin_return_address(0), &rpath, &runpath);
 
+  // 0. If filename is null, call libc dlopen. The user just wants a handle to
+  // the main program.
+  if (filename == NULL) {
+    ret = _real_dlopen(filename, flag);
+  }
+
   // 1. Attempt dlopen using RPATH.
-  if (!rpath.empty()) {
+  if (ret == NULL && !rpath.empty()) {
     ret = _real_dlopen(filename, flag);
```

Comment:
I suggest:
> `if filename == NULL || !rpath.empty()) {`

on the above line, and then remove the changes at 149 to 152.

In summary, the new code should look like the following.  This avoids duplicating the logic of the call to `_real_dlopen()`:
```
// 1a. If filename is NULL, call libc dlopen. The user just wants a handle to // the main program.
// 1b. If RPATH exists, call libc dlopen using RPATH.
if filename == NULL || !rpath.empty()) {`
  ret = _real_dlopen(filename, flag); }
```